### PR TITLE
3. Extended stats feature

### PR DESCRIPTION
**Pull request by kittyevi**
PR: #3

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -9,7 +9,7 @@
   <body>
     <div class="welcome-container">
       <h1 class="welcome-container__headline">Welcome</h1>
-      <p class="welcome-container__paragraph">TODO: General description of how the website works</p>
+      <p class="welcome-container__paragraph"></p>
```
Comment (stefanluth) [inline index.html]: The best way to handle TODOs is to remove them :D

**Review by MKrips — APPROVED** [review]

File: script_files/displayText.js [hunk 2]
```diff
@@ -18,7 +18,7 @@ export function cleanText(text) {
 }
 
 export function charArrayIntoString(textArray) {
-  var resultString = "";
+  let resultString = "";
   for (const char of textArray) {
     resultString += char;
   }
```
Comment (gazer42) [inline script_files/displayText.js]: use the join method for this

File: script_files/main.js [hunk 3]
```diff
@@ -7,24 +7,37 @@ import {incrementSeconds} from "./timeHandler.js";
...
-
 let charArray = [];
 let charIndex = 0;
 let seconds = 0;
 let userKeyTypeCount = 0;
 let userMistakesCount = 0;
 let localStorage = window.localStorage
+let extendedStatsString = "";
+
+keyboardKeysArray.forEach(function(element) {
```
Comment (gazer42) [inline script_files/main.js]: element is not a good name

File: script_files/main.js [hunk 4]
```diff
@@ -7,24 +7,37 @@ import {incrementSeconds} from "./timeHandler.js";
...
+  extendedStatsString += element[0] + ' ' + element[1] + '%' + '<br>';
```
Comment (gazer42) [inline script_files/main.js]: a template string would be easier to read

File: script_files/main.js [hunk 5]
```diff
@@ -107,5 +125,45 @@ function resetProgress() {
...
+    seconds = incrementSeconds(seconds, statsTextForSeconds, todayStatsText, totalStatsText);
+    let minutes = seconds / 60;
+    const charactersPerMinute = userKeyTypeCount / minutes;
+    const isFreshRun = userKeyTypeCount == 0;
+
+    if (isFreshRun) {
+      statsText.innerHTML = 'CPM: ' + '0 ' + 'Wrong Chars: ' + '0%';
+    }
+    else {
+      statsText.innerHTML = 'CPM: '+ Math.round(charactersPerMinute) + ' ' + 'Wrong Chars: ' + Math.round((userMistakesCount * 100 / userKeyTypeCount * 100) / 100) + '%';
```
Comment (gazer42) [inline script_files/main.js]: I think template strings would look much cleaner

**Review by gazer42 — CHANGES_REQUESTED** [review]

File: index.html (hunk 1 above)
Comment (kittyevi) [reply index.html]: Lol

File: script_files/main.js (hunk 3 above)
Comment (kittyevi) [reply script_files/main.js]: 😂😂

File: index.html (hunk 1 above)
Comment (stefanluth) [inline index.html]: You can probably remove this if you don't plan on writing a description :D

File: script_files/displayStats.js [hunk 6]
```diff
@@ -1,12 +1,12 @@
 export function displayStats(userMistakesCount, userKeyTypeCount, seconds) {
   if (userMistakesCount == 0 && userKeyTypeCount == 0) {
-    return 'CPM: ' + '0 ' + 'Wrong Chars: ' + '0%';
+    return `CPM: 0 Wrong Chars: 0%`;
```
Comment (stefanluth) [inline script_files/displayStats.js]:
```suggestion
    return 'CPM: 0 Wrong Chars: 0%';
```

File: script_files/displayStats.js [hunk 7]
```diff
@@ -17,6 +17,15 @@ export function displayTotalStats() {
   return putTogetherStringsForStats('totalTotalSeconds', 'totalCharsTyped', 'totalMistypes', 'totalAmountOfSets');
 }
 
+export function showExtendedStats(extendedStatsContainer) {
```
Comment (stefanluth) [inline script_files/displayStats.js]:
```suggestion
export function toggleExtendedStats(extendedStatsContainer) {
```

File: script_files/displayText.js [hunk 8]
```diff
@@ -16,13 +16,18 @@ export function cleanText(text) {
   return resultText;
 }
 
+function cleanTextWithoutCommaWhitespace(text) {
+  const resultText = text
+    .replaceAll(',', "");
```
Comment (stefanluth) [inline script_files/displayText.js]:
```suggestion
  const resultText = text.replaceAll(',', "");
```

File: script_files/displayText.js [hunk 9]
```diff
@@ -16,13 +16,18 @@ export function cleanText(text) {
   return resultText;
 }
 
+function cleanTextWithoutCommaWhitespace(text) {
```
Comment (stefanluth) [inline script_files/displayText.js]:
```suggestion
function removeWhitespaces(text) {
```

File: script_files/displayText.js [hunk 10]
```diff
@@ -16,13 +16,18 @@ export function cleanText(text) {
...
+
+  return resultText
+}
+
 export function charArrayIntoString(textArray) {
-  var resultString = "";
-  for (const char of textArray) {
-    resultString += char;
-  }
+  let resultString = textArray.join();
```
Comment (stefanluth) [inline script_files/displayText.js]:
```suggestion
  const resultString = textArray.join();
```

File: script_files/main.js [hunk 11]
```diff
@@ -45,7 +59,7 @@ generateTextButton.addEventListener('click', async function() {
     const charactersPerMinute = userKeyTypeCount / minutes;
 
     if (userKeyTypeCount == 0) {
-      statsText.innerHTML = 'CPM: ' + '0 ' + 'Wrong Chars: ' + '0%';
+      statsText.innerHTML = `CPM: 0, Wrong Chars: 0%`;
```
Comment (stefanluth) [inline script_files/main.js]:
```suggestion
      statsText.innerHTML = 'CPM: 0, Wrong Chars: 0%';
```

File: script_files/main.js [hunk 12]
```diff
@@ -1,36 +1,50 @@
...
 localStorage = parseLocalStorage(localStorage);
 
 let generateTextButtonIsClicked = false;
-statsTextForSeconds.innerHTML = 'Seconds: ' + seconds + 's';
+statsTextForSeconds.innerHTML = `Seconds: ${seconds}s`;
 statsText.innerHTML = displayStats(0, 0);
 todayStatsTextContainer.innerHTML = displayTodayStats();
 totalStatsTextContainer.innerHTML = displayTotalStats();
 
+let firstTryCounter = 0;
```
Comment (stefanluth) [inline script_files/main.js]:
```suggestion
let tryCounter = 0;
```

**Review by stefanluth — CHANGES_REQUESTED** [review]
Hitting any key, including Shift, Esc, Arrow Keys, etc. increases the chars typed counter. The timer continues counting even after finishing typing a set :)

File: index.html (hunk 1 above)
Comment (kittyevi) [reply index.html]: Fixed it in the branch :D May it be that you didn't switch on my branch by mistake? 😁

**Comment by kittyevi** [comment]
> Hitting any key, including Shift, Esc, Arrow Keys, etc. increases the chars typed counter. The timer continues counting even after finishing typing a set :)

I didn't prevent the shift, esc, arrow keys etc. from default on purpose because in my opinion these keys should add to the key type counter because these keys are not the correctly typed key that should be pressed. Additionally the timer of the current set resets but the timer for whole days and so forth should continue, so it works as I expected. However, if you want me to change things accordingly I will do it with pleasure. :)

**Comment by stefanluth** [comment]
> I didn't prevent the shift, esc, arrow keys etc. from default on purpose because in my opinion these keys should add to the key type counter because these keys are not the correctly typed key that should be pressed. Additionally the timer of the current set resets but the timer for whole days and so forth should continue, so it works as I expected. However, if you want me to change things accordingly I will do it with pleasure. :)

I would expect the `Today` and `Total` stats to only be the time I have spent typing today/in total and for the time to only start after I have started typing and to stop when I finished a given text :)

**Comment by kittyevi** [comment]
> 

Hence the "Today" Stats work as expected via resetting at midnight and the "Total" stats working as well, I will concentrate my work now on the timer being stopped when finishing a given text AND continueing when beginning to type again... And this is a bit trickier as expected because I don't know how to solve it with the setInterval function. I will try though :)

**Comment by kittyevi** [comment]
> 

Fixed it, finally 😁

**Comment by stefanluth** [comment]
The timer starts before I start typing and I can't generate another text after generating one :(

https://user-images.githubusercontent.com/80166473/212631917-026c973f-237b-4ef8-acee-42f0316dc135.mov

**Comment by kittyevi** [comment]
> The timer starts before I start typing and I can't generate another text after generating one :(
> 
>  Screen.Recording.2023-01-16.at.09.28.13.mov

May I ask you to take a look now? :) Is it fixed as you expected?

**Comment by stefanluth** [comment]
Yesss, works now. :) Still a few issues:

The current letter is colored red, even though I didn't mistype. Hitting "S" once registers as two. Maybe also the reason for the red indicator?

https://user-images.githubusercontent.com/80166473/212673102-cb5dbf73-532a-44db-a4f9-5d3f307bc540.mov

Edit: To reproduce, generate a few texts and type them.
Edit2: It seems like each text generation adds another event listener.

**Comment by kittyevi** [comment]
> 

Hopefully, it SHOULD be fixed by now 😂😁 May you take a look?

**Comment by stefanluth** [comment]
Works! All that's left to change:

image

:)

**Comment by kittyevi** [comment]
> Works! All that's left to change:
> 
> image
> 
> :)

Oh, how did this occur? :o

**Comment by stefanluth** [comment]
As soon as I open the page :)

File: index.html (hunk 1 above)
Comment (stefanluth) [reply index.html]: Which branch? The branch of this PR still has the empty paragraph :)

Comment (kittyevi) [reply index.html]: lol

**Review by stefanluth — APPROVED** [review]

**Pull request merged**